### PR TITLE
Refactor config.json into sku model and cleanup error messages

### PR DESCRIPTION
**Pull request by zyoung51**
- Change InternalError errors to BadRequest when the issue arises due to
  bad data in the skupack file that is posted.
- Move the config.json properties into the model

@RackHD/corecommitters @uppalk1 @keedya @tannoa2

### CODE REVIEW

**Comment by zyoung51** [comment]
test this please

File: lib/services/sku-pack-service.js [hunk 1]
```diff
@@ -444,28 +462,30 @@ function skuPackServiceFactory(
      * @param  {String}     fromRoot The path to the configuration file
      * @return {Promise}
      */
-    SkuPackService.prototype.validatePack = function(contents, fromRoot) {
-        try {
+    SkuPackService.prototype.validatePack = function(contents, fromRoot, options) {
+        options = options || {};
+
+        return Promise.try(function() {
             var conf = JSON.parse(contents);
```
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Could you please add a try/catch for JSON.parse? so the invalid JSON format error can be explicitly threw out to user,  I had said this was an enhance opportunity in this issue https://github.com/RackHD/RackHD/issues/190

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: Promise.try() converts synchronous throws into rejects to be caught by .catch().  The error message, along with the notice about the invalid json, occurs at line 487.

File: lib/services/sku-pack-service.js [hunk 2]
```diff
@@ -477,63 +497,69 @@ function skuPackServiceFactory(
...
-                        throw new Error('rules or name is missing');
+        .then(function(fileContents) {
+            contents = fileContents;
+            return self.validatePack(fileContents, fromRoot, {
+                validateSku: !_.isUndefined(skuid)
+            });
+        })
+        .then(function() {
+            if(skuid === undefined)  {
+                var conf = JSON.parse(contents);
```
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Also to add a try/catch for this JSON.parse

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: Pre-validated in validatePack above.

File: lib/services/sku-pack-service.js [hunk 3]
```diff
@@ -477,63 +497,69 @@ function skuPackServiceFactory(
...
-                return [skuID, fs.readdirAsync(fromRoot), fs.mkdirAsync(dst) ];
-            })
-            .spread(function(sku, entries) {
-                return Promise.map(entries, function(entry) {
-                    var src = fromRoot + '/' + entry;
-                    if( entry === 'config.json' ) {
-                        return fs.renameAsync(src, dst + '.json');
-                    } else {
-                        return fs.renameAsync(src, dst + '/' + entry);
+                    return waterline.skus.create(_.pick(conf, ['name', 'rules', 'discoveryGraphName', 'discoveryGraphOptions'] ));
```
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Why only pick these keys? From your SKU model definition in on-core (https://github.com/RackHD/on-core/pull/123/files) , other keys such as "workflowRoot", "taskRoot" should also be inserted.

So either PR need a fix.

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: Line 542...

File: lib/services/sku-pack-service.js [hunk 4]
```diff
@@ -477,63 +497,69 @@ function skuPackServiceFactory(
...
-                    if(!conf.hasOwnProperty('rules') || !conf.hasOwnProperty('name')) {
-                        throw new Error('rules or name is missing');
+        .then(function(fileContents) {
+            contents = fileContents;
+            return self.validatePack(fileContents, fromRoot, {
+                validateSku: !_.isUndefined(skuid)
+            });
+        })
+        .then(function() {
+            if(skuid === undefined)  {
```
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: `undefined` seems not enough, other values like `null` need be judged as well.

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: The controller code validates that the skuid exists if it is defined.  It is already validated.

File: lib/services/sku-pack-service.js (hunk 1 above)
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Oh, yes, I see it. Thanks.

File: lib/services/sku-pack-service.js (hunk 3 above)
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Why split to two db operations? Any reason to design like this?

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: In the case where we are not creating the sku (skuid is defined), we are only modifying the sku pack directories and not replacing the sku rules/names

**Comment by jlongever** [comment]
test this please

File: lib/services/sku-pack-service.js (hunk 3 above)
Comment (yyscamper) [inline lib/services/sku-pack-service.js]: Sorry, still confused, not creating the SKU? but this is `waterline.skus.create`, it is sure to create a new SKU.

Comment (zyoung51) [inline lib/services/sku-pack-service.js]: It creates a new sku if this was called with the sku undefined as in POST /skus/pack.  It does not create or update the sku related model settings if it was called with the sku defined when replacing the pack binaries as in PUT /skus/:id/pack.  In both cases the .update is called to create/replace the skupack related model settings.

**Pull request merged**